### DISCUSSION TITLE
Literal and case-insensivity?

### DISCUSSION THREAD

**Discussion by pawamoy** [DanCardin/cappa#161] [Q&A] [answered]
I'm trying to allow both `debug` and `DEBUG` as input value of this option:

```python
@dataclass(kw_only=True)
class LogLevelPathOptions:
    """Log level and path options."""

    log_level: An[
        Literal["TRACE", "DEBUG", "INFO", "SUCCESS", "WARNING", "ERROR", "CRITICAL"],
        cappa.Arg(short="-L", long=True, parse=str.upper),
        Doc("Log level to use when logging messages."),
    ] = "INFO"
```

If I read [the docs](https://cappa.readthedocs.io/en/latest/annotation.html#literals-and-enums) correctly, `Literal` implies `choices=...` which seems to skip my `parse` callable, resulting in the following error:

```
pypi-insiders: error: Invalid choice: 'debug' (choose from 'TRACE', 'DEBUG', 'INFO', 'SUCCESS', 'WARNING', 'ERROR', 'CRITICAL')
```

How would I go about achieving that? Should I store the literal in a type variable and use `MyLiteral.__args__` to build the choices? I think it would print both upper and lowercase choices in the help, which wouldn't be great.

**Comment by DanCardin** [comment]
mm so `choices` are currently being enforced at the parser level, while all parsing is done after the parsing has already completed.

There **is** also parsing-level validation of literals, but as a general pattern `Arg(parse=...)` completely replaces any inference based parsing.

With that said, in this scenario, since it doesn't otherwise affect the parsed value, it's just validation on top of it, I could see where it could make sense to avoid checking choices inside the parser at all, and to only do so at the parsing/mapping phase.

**Reply by pawamoy** [reply]
To be clear, having the error message mentioned above is great:

> pypi-insiders: error: Invalid choice: 'something' (choose from 'TRACE', 'DEBUG', 'INFO', 'SUCCESS', 'WARNING', 'ERROR', 'CRITICAL')

...and I suppose it's thanks to `choices`, so it's great that `choices` is automatically populated from the literal :+1: 

I simply expected my `parse` callable to perform a transformation *before* the value is checked against valid choices :relaxed: But not knowing Cappa's internals, I'm not sure how much sense it makes or how difficult it is to change.

I trust your final decision anyway!

**Reply by DanCardin** [reply]
I'm looking into whether I can easily remove choices validation from the parser (and thus argparse when using the argparse backend, which is ideal) and deferring it into the parsing phase. It's probably doable, and likely a good idea if so.

I also did a quick test with just argaparse

```python
import argparse
p = argparse.ArgumentParser()
p.add_argument('-p', choices=['A', 'B', 'C'], type=str.upper)
print(p.parse_args(['-p', 'a']))
```

which **does** work, so it's also a positive indicator.

---

Relatedly, I assume you'd expect `Literal[1, 2, 3]` for `'asdf'` input to produce `invalid choice: 'asdf' (choose from 1, 2, 3)` rather than `invalid literal for int() with base 10: 'asdf'` (I think i do). Which complicates things somewhat)

**Reply by pawamoy** [reply]
Good point. I suppose I'd prefer seeing the raw value indeed (the thing the user typed in the console). There might be cases where I would *also* want to see errors coming from value transformations (like `int()`-ing a non-digit string, as you shown), but I don't have any good example right now :thinking: So maybe go "simple" first, and just show the raw value :+1: 

If it proves too hard, it's also fine to just see the last error caused by a transformation, or the last transformed value. In my case it would show

```
pypi-insiders: error: Invalid choice: 'FATAL' (choose from 'TRACE', 'DEBUG', 'INFO', 'SUCCESS', 'WARNING', 'ERROR', 'CRITICAL')
```

...instead of

```
pypi-insiders: error: Invalid choice: 'fatal' (choose from 'TRACE', 'DEBUG', 'INFO', 'SUCCESS', 'WARNING', 'ERROR', 'CRITICAL')
```

...if the user happened to type it lowercase. In this case it's completely fine to show the transformed value, as it kinda indicates that the value is uppercase'd, which is good for the user to know.

**Reply by DanCardin** [reply]
I guess you're explicitly disabling completions, but that'd be a reason why choices are relevant to the parser either way. Not that the parser has any specific need to raise versus during the mapping phase.

I think it should be doable, i'm just struggling to find the right way to formulate it with error handling and sequences of literals and explicit parsers and such.

I anticipate that `invalid choice: 'FATAL'` is going to be the easier route error-message wise.

**Reply by pawamoy** [reply]
> I guess you're explicitly disabling completions, but that'd be a reason why choices are relevant to the parser either way. Not that the parser has any specific need to raise versus during the mapping phase.

Sorry, it's unclear from the code I shared, but I declare the `completion` option myself, to reword its help message, and that's why I'm disabling it the default option in the main `invoke` call.

> I think it should be doable, i'm just struggling to find the right way to formulate it with error handling and sequences of literals and explicit parsers and such.
I anticipate that invalid choice: 'FATAL' is going to be the easier route error-message wise.

That's a fair trade-off, as the valid options are shown in the help message anyway :slightly_smiling_face:

**Reply by pawamoy** [reply]
If you make this change, and I add back `choices` myself, will my `parse` callable still work as expected? Because I'm happy to manually provide the choices if they can't be inferred anymore, for error message purposes.

**Reply by DanCardin** [reply]
Fwiw, you can do `completion=Arg(...)` to define what the argument should look like.

If this works, I expect the annotation to still imply choices, just non-choice choices wont error inside the parser (and i wont even supply them to argparse, because then **it** will error). then the mapping will do the check and fail there, so it shouldnt require any changes to the original impl you supplied, i wouldnt think.

**Reply by DanCardin** [reply, marked as answer]
https://github.com/DanCardin/cappa/pull/179
